### PR TITLE
std::isnan, std::isfinite, std::unique_ptr, std::shared_ptr

### PR DESCRIPTION
**Pull request by pbauman**
Now, C++11 is turned on by default in the compiler. Adds infrastructure to test for, and use if detected, std::isnan, std::isfinite, std::unique_ptr, and std::shared_ptr; these tests were inspired by or straight up copied from libMesh/cxx11.m4. We fall back to Boost if not supported. @roystgnr note that I had to turn off the C++11 style asserts in 5cb0940 to get QUESO to compile with C++11. There was a cascading set of failures to convert types. I tried to replace all the places we use Boost with the macros/typedefs, except for program options. So one step in the direction of making Boost optional. @dmcdougall I didn't have a SLEPc-enabled libMesh lying around, so I may have missed something in there. Could you please check that I didn't hose that?

I'm not sure we want to merge this right away. I like moving farther away from the boost requirement, but program_options uses boost::shared_ptr under the hood and so there are LOTS of warnings emitted about std::auto_ptr deprecation (since boost::shared_ptr provides an interface to std::auto_ptr). Those warnings definitely are verbose and could hide legit warnings.

### CODE REVIEW

**Comment by roystgnr** [comment]
> I had to turn off the C++11 style asserts in 5cb0940 to get QUESO to compile with C++11.

What was the compilation failure otherwise?  I suppose the C++11 case might not have been getting much use here, but those asserts are pretty much cut and pasted from https://github.com/libantioch/antioch/blob/master/src/utilities/include/antioch/antioch_asserts.h#L63 and we use C++11 there regularly.

File: src/core/src/InfiniteDimensionalMCMCSampler.C [hunk 1]
```diff
@@ -27,7 +27,6 @@
 
 #include <cmath>
 #include <algorithm>
```
Comment (dmcdougall) [inline src/core/src/InfiniteDimensionalMCMCSampler.C]: Can you add `#include <sstream>` here?  Getting a compiler error regarding incomplete `stringstream` types below.

**Comment by dmcdougall** [comment]
>> I had to turn off the C++11 style asserts in 5cb0940 to get QUESO to compile with C++11.
>
> What was the compilation failure otherwise? I suppose the C++11 case might not have been getting much use here, but those asserts are pretty much cut and pasted from https://github.com/libantioch/antioch/blob/master/src/utilities/include/antioch/antioch_asserts.h#L63 and we use C++11 there regularly.

I get this one:

```
In file included from ../inc/queso/Defines.h:29:0,
                 from ../inc/queso/Environment.h:28,
                 from core/src/Environment.C:28:
core/src/Environment.C: In member function ‘void QUESO::BaseEnvironment::scanInputFileForMyOptions(const boost::program_options::options_description&) const’:
../inc/queso/asserts.h:76:196: error: invalid static_cast from type ‘const string {aka const std::__cxx11::basic_string<char>}’ to type ‘type2 {aka const char (&)[1]}’
 fine queso_require_not_equal_to_msg(expr1,expr2,msg)  do { typedef decltype(expr1) type1; typedef decltype(expr2) type2; if (!((expr1 != static_cast<type1>(expr2)) && (static_cast<type2>(expr1) != expr2)
                                                                                                                                                                                                 ^
core/src/Environment.C:372:3: note: in expansion of macro ‘queso_require_not_equal_to_msg’
   queso_require_not_equal_to_msg(m_optionsInputFileName, "", "m_optionsInputFileName is 'nothing'");
```

File: src/core/src/InfiniteDimensionalMCMCSampler.C (hunk 1 above)
Comment (pbauman) [inline src/core/src/InfiniteDimensionalMCMCSampler.C]: da fuq, what compiler are you using?

**Comment by dmcdougall** [comment]
Replacing `""` with `((std::string)(""))` appears to fix the comparison with an empty `const char *`.

File: src/core/src/InfiniteDimensionalMCMCSampler.C (hunk 1 above)
Comment (pbauman) [inline src/core/src/InfiniteDimensionalMCMCSampler.C]: Re: @roystgnr here's an example, but they're all type-conversion errors: 
<pre>
In file included from ../inc/queso/Defines.h:29:0,
                 from ../inc/queso/Environment.h:28,
                 from ../inc/queso/Miscellaneous.h:28,
                 from ../../../pbauman/src/core/src/EnvironmentOptions.C:28:
../../../pbauman/src/core/src/EnvironmentOptions.C: In constructor ‘QUESO::EnvironmentOptions::EnvironmentOptions(const QUESO::BaseEnvironment&, const char*)’:
../inc/queso/asserts.h:76:196: error: invalid static_cast from type ‘std::__cxx11::string {aka std::__cxx11::basic_string<char>}’ to type ‘type2 {aka const char (&)[1]}’
 #define queso_require_not_equal_to_msg(expr1,expr2,msg)  do { typedef decltype(expr1) type1; typedef decltype(expr2) type2; if (!((expr1 != static_cast<type1>(expr2)) && (static_cast<type2>(expr1) != expr2))) { std::cerr << "Assertion `" #expr1 " != " #expr2 "' failed.\n" #expr1 " = " << (expr1) << "\n" #expr2 " = " << (expr2) << '\n' << msg << std::endl; queso_error(); } } while(0)
                                                                                                                                                                                                    ^
../../../pbauman/src/core/src/EnvironmentOptions.C:255:3: note: in expansion of macro ‘queso_require_not_equal_to_msg’
   queso_require_not_equal_to_msg(m_env.optionsInputFileName(), "", "this constructor is incompatible with the abscense of an options input file");
</pre>

A lot of them could be fixed with making things like `""` or `"="` explicitly `std::string("")` etc (and I actually started sed'ing for those), but then there were others and I lost patience:

<pre>
In file included from ../inc/queso/Defines.h:29:0,
                 from ../inc/queso/DistArray.h:28,
                 from ../inc/queso/VectorSpace.h:28,
                 from ../inc/queso/VectorSequence.h:30,
                 from ../../../pbauman/src/basic/src/VectorSequence.C:26:
../../../pbauman/src/basic/src/VectorSequence.C: In instantiation of ‘double QUESO::BaseVectorSequence<V, M>::unifiedPositionsOfMaximum(const QUESO::ScalarSequence<double>&, QUESO::BaseVectorSequence<V, M>&) [with V = QUESO::GslVector; M = QUESO::GslMatrix]’:
../../../pbauman/src/basic/src/VectorSequence.C:2910:23:   required from here
../inc/queso/asserts.h:75:134: error: invalid static_cast from type ‘int’ to type ‘type1 {aka int&}’
 #define queso_require_equal_to_msg(expr1,expr2,msg)  do { typedef decltype(expr1) type1; typedef decltype(expr2) type2; if (!((expr1 == static_cast<type1>(expr2)) && static_cast<type2>(expr1) == expr2)) { std::cerr << "Assertion `" #expr1 " == " #expr2 "' failed.\n" #expr1 " = " << (expr1) << "\n" #expr2 " = " << (expr2) << '\n' << msg << std::endl; queso_error(); } } while(0)
                                                                                                                                      ^
../../../pbauman/src/basic/src/VectorSequence.C:495:5: note: in expansion of macro ‘queso_require_equal_to_msg’
     queso_require_equal_to_msg(recvcntsPos[0], (int) subNumPos, "failed MPI.Gather() result at proc 0 (recvcntsPos[0])");
</pre>

**Comment by pbauman** [comment]
> Can you add #include <sstream> here? Getting a compiler error regarding incomplete stringstream types below.

> da fuq, what compiler are you using?

Oh right, that file wasn't getting hit because of lack of SLEPc. OK, I'll drop the header in and squash it.

**Comment by pbauman** [comment]
> Oh right, that file wasn't getting hit because of lack of SLEPc. OK, I'll drop the header in and squash it.

Done.

File: src/misc/inc/math_macros.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+namespace QUESO
+{
+#ifdef QUESO_HAVE_CXX11_ISNAN
+#define QUESO_ISNAN(arg) std::isnan(arg)
+#elif QUESO_HAVE_BOOST_MATH_SPECIAL_FUNCTIONS_HPP
+#define QUESO_ISNAN(arg) (boost::math::isnan)(arg)
+#else
+#     error "No valid definition for is_nan found!"
+#endif
+
```
Comment (roystgnr) [inline src/misc/inc/math_macros.h]: Why make these macros rather than template functions?

Comment (pbauman) [inline src/misc/inc/math_macros.h]: No good reason. I can do template functions instead if you prefer.

**Comment by dmcdougall** [comment]
@pbauman This needs to be updated to reflect the changes made in #441.

File: src/misc/inc/math_macros.h (hunk 2 above)
Comment (dmcdougall) [inline src/misc/inc/math_macros.h]: Is there a big design difference between the two?  Internally, boost's `isnan` is a template function.

Comment (roystgnr) [inline src/misc/inc/math_macros.h]: Let's do that.  I can't imagine needing to debug inside isnan() or isfinite(), but "use macros only when there's no alternative" is a good habit to reinforce.  Also then we've got an excuse to use uncapitalized function names and make the code look less shouty.

**Comment by roystgnr** [comment]
Rebasing, macro->template_function change, and GPMSA.C isnan replacement (use isfinite there?  I should have picked that instead), and this should be ready to merge.  Sorry I didn't review sooner.

**Comment by dmcdougall** [comment]
I disagree, we're getting compiler errors with comparisons to `""` in our assert macros.  We'll have to convert those to equivalent `std::string` objects before merging.

File: src/misc/inc/math_macros.h (hunk 2 above)
Comment (dmcdougall) [inline src/misc/inc/math_macros.h]: > "use macros only when there's no alternative" 

Ok, I'm happy with this, and I agree with the overall plan.  Thanks.

**Comment by pbauman** [comment]
> we're getting compiler errors with comparisons to "" in our assert macros. We'll have to convert those to equivalent std::string objects before merging.

That's if you exclude 5cb0940 in this PR. Not saying we shouldn't exclude it, but just to be clear, this PR compiles and `make check` passes as-is (without the aforementioned updates needed) throughout the history.

Also, it's more than just the `std::string` bits. Once you knock those out, there's more conversion problems that pop up and I never finished going through them because I got tired of playing whack-a-mole.

**Comment by dmcdougall** [comment]
Ok, good to know, and thanks for exploring that.  Hopefully you didn't waste too much time down that rabbit hole.  I suppose most of our asserts are usually on `int`s, `double`s or `const char *`s.  We'd get weird behaviour if for some reason the compiler converted to the wrong type for objects.

**Comment by roystgnr** [comment]
We need to fix those macros at some point, though.  My attempt at robustness via overly-complicated overly-restrictive casting was wrong in multiple senses, but the simple macros have scary failure modes too.

Let me explain via demonstration?

```
roystgnr@minerva:~
$ cat test.C
#include <iostream>

int main (void)
{
  unsigned int i = 100;
  if (i < -5)
    std::cout << "100 < -5" << std::endl;
  else
    std::cout << "100 > -5" << std::endl;

  return 0;
}
roystgnr@minerva:~
$ g++ test.C -o test && ./test
100 < -5
```

**Comment by dmcdougall** [comment]
Yikes, yeah.  That's a really good point.  There are definitely places where we compare signed and unsigned ints in these macros, too.

**Comment by pbauman** [comment]
OK, this is almost done. I've rebased on dev and I redid the C++11 assert macros in the same vein as libantioch/antioch#213. Everything is working for me on GCC 6.1. Went ahead and pushed so Travis can have a go. Only remaining item is to remove the macros in-lieu of template functions for `isnan` and `isinfinite`.

**Comment by pbauman** [comment]
Weird. Regenerated headers and trying again.

**Comment by pbauman** [comment]
OK, there's some header include dependency crap going on between compiler versions. Let me build 4.6.3 locally and see if I can reproduce.

**Comment by pbauman** [comment]
Third time's the charm? I could reproduce the failure with GCC 4.6.3 on my workstation. Now compiles/make checks successfully for me. I also converted the macros to template functions so, if Travis is happy, this should be good to go.

Again, we'll probably want this to go in closely with the optional program options (#451) so we can opt out of Boost altogether. Otherwise there is just a ton of warnings from modern GCC versions about the deprecated `std::auto_ptr` in the `boost::shared_ptr` interface.

**Comment by dmcdougall** [comment]
Looks good.  I was looking at your older commits yesterday and started some of the work.  I woke up this morning to a bunch of your GitHub comments and completed work.  Our solutions were basically identical.  My local checks are happy, and Travis is happy too, so I'm going to merge this.

**Comment by dmcdougall** [comment]
By the way, we ought to figure a way to have Travis test multiple versions of gcc and clang.

**Comment by dmcdougall** [comment]
Ah, you forgot some `isnan`s in GPMSA.C.  I'll get those.

**Comment by dmcdougall** [comment]
@pbauman Passing `--disable-boost-program-options` from #451, I still see warnings triggered by [this line](https://github.com/libqueso/queso/pull/430/files#diff-885a0e344e635932ea95e05c68004329R33).

Here's an example warning:

```
CXX      stats/src/GaussianJointPdf.lo
In file included from /opt/apps/ossw/libraries/boost/boost-1.58.0/c7/gcc-5.2/include/boost/smart_ptr/shared_ptr.hpp:28:0,
                 from /opt/apps/ossw/libraries/boost/boost-1.58.0/c7/gcc-5.2/include/boost/shared_ptr.hpp:17,
                 from /opt/apps/ossw/libraries/boost/boost-1.58.0/c7/gcc-5.2/include/boost/format/alt_sstream.hpp:21,
                 from /opt/apps/ossw/libraries/boost/boost-1.58.0/c7/gcc-5.2/include/boost/format/internals.hpp:23,
                 from /opt/apps/ossw/libraries/boost/boost-1.58.0/c7/gcc-5.2/include/boost/format.hpp:38,
                 from /opt/apps/ossw/libraries/boost/boost-1.58.0/c7/gcc-5.2/include/boost/math/policies/error_handling.hpp:31,
                 from /opt/apps/ossw/libraries/boost/boost-1.58.0/c7/gcc-5.2/include/boost/math/special_functions/gamma.hpp:23,
                 from /opt/apps/ossw/libraries/boost/boost-1.58.0/c7/gcc-5.2/include/boost/math/special_functions/detail/bessel_jy.hpp:14,
                 from /opt/apps/ossw/libraries/boost/boost-1.58.0/c7/gcc-5.2/include/boost/math/special_functions/bessel.hpp:20,
                 from /opt/apps/ossw/libraries/boost/boost-1.58.0/c7/gcc-5.2/include/boost/math/special_functions/airy.hpp:12,
                 from /opt/apps/ossw/libraries/boost/boost-1.58.0/c7/gcc-5.2/include/boost/math/special_functions.hpp:15,
                 from ../inc/queso/math_macros.h:33,
                 from ../inc/queso/JointPdf.h:33,
                 from ../inc/queso/GaussianJointPdf.h:30,
                 from ../../src/stats/src/GaussianJointPdf.C:25:
/opt/apps/ossw/libraries/boost/boost-1.58.0/c7/gcc-5.2/include/boost/smart_ptr/detail/shared_count.hpp:395:33: warning: ‘template<class> class std::auto_ptr’ is deprecated [-Wdeprecated-declarations]
     explicit shared_count( std::auto_ptr<Y> & r ): pi_( new sp_counted_impl_p<Y>( r.get() ) )
```

I'm trying to figure out how it's still seeing that header.  If I comment out that line, I don't get any warnings.

**Pull request merged**